signing/signing.go

```diff
@@ -1,22 +1,390 @@
+	}
+
+	// Remove repo path from workflow filepath.
+	ind := strings.Index(workflowRef, repoPath) + len(repoPath) + 1
+	workflowPath = workflowRef[ind:]
+
+	// Remove repo ref tag from workflow filepath.
+	workflowPath = workflowPath[:strings.Index(workflowPath, "@")]
+
+	return repoPath, repoRef, repoSHA, workflowPath, nil
+}
+
+func verifyScorecardWorkflow(workflowContent string) error {
+	// Verify workflow contents using actionlint.
+	workflow, lintErrs := actionlint.Parse([]byte(workflowContent))
+	if lintErrs != nil {
+		return fmt.Errorf("actionlint errors parsing workflow: %v", lintErrs)
+	}
+
+	// Verify that there are no global env vars or defaults.
+	if workflow.Env != nil || workflow.Defaults != nil {
+		return errors.New("workflow contains global env vars or defaults")
+	}
+
+	// Verify that the all scope, if set, isn't write-all.
+	globalPermAll := workflow.Permissions.All
+	if globalPermAll != nil && globalPermAll.Value == "write-all" {
+		return fmt.Errorf("global perm is set to write-all")
+	}
+
+	// Verify that there are no global permissions (including id-token) set to write.
+	globalPerms := workflow.Permissions.Scopes
+	for globalPerm, val := range globalPerms {
+		if val.Value.Value == "write" {
+			return fmt.Errorf("global perm %v is set to write", globalPerm)
+		}
+	}
+
+	// Find the (first) job with a step that calls scorecard-action.
+	scorecardJob := findScorecardJob(workflow.Jobs)
+	if scorecardJob == nil {
+		return errors.New("workflow has no job that calls ossf/scorecard-action")
+	}
+
+	// Make sure other jobs don't have id-token permissions.
+	for _, job := range workflow.Jobs {
+		if job != scorecardJob && job.Permissions != nil {
+			idToken := job.Permissions.Scopes["id-token"]
+			if idToken != nil && idToken.Value.Value == "write" {
+				return errors.New("workflow has a non-scorecard job with id-token permissions")
+			}
+		}
+	}
+
+	if scorecardJob == nil {
+		return errors.New("workflow has no job calling ossf/scorecard-action")
+	}
+
+	// Verify that there is no job container or services.
+	if scorecardJob.Container != nil || len(scorecardJob.Services) > 0 {
+		return errors.New("workflow contains container or service")
+	}
+
+	// Verify that the workflow runs on ubuntu and nothing else.
+	if scorecardJob.RunsOn == nil {
+		return errors.New("no RunsOn found in workflow")
+	} else {
+		labels := scorecardJob.RunsOn.Labels
+		if len(labels) != 1 {
+			return errors.New("workflow doesn't have only 1 virtual environment")
+		}
+		jobEnv := labels[0].Value
+		if jobEnv != "ubuntu-latest" && jobEnv != "ubuntu-20.04" && jobEnv != "ubuntu-18.04" {
+			return errors.New("workflow doesn't run on ubuntu")
+		}
+	}
+
+	// Verify that there are no job env vars set.
+	if scorecardJob.Env != nil {
+		return errors.New("workflow contains env vars")
+	}
+
+	// Verify that there are no job defaults set.
+	if scorecardJob.Defaults != nil {
+		return errors.New("workflow has defaults set")
+	}
```

Comment:
I think you also need to check for the absence of `container` https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idcontainer
and `service` https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idservices

because those can be used to execute user-defined code.

Reply:
This is already done further up in the code (line 268)

Reply:
ah my bad. I was under the impression those could also be used at the top-level of the workflow. I'm paranoid :-)